sdk/eventhub/azure-eventhubs/azure/eventhub/_producer_client.py

```diff
@@ -0,0 +1,170 @@
+    receiving events from the Azure Event Hubs service.
+
+    :param str host: The hostname of the Event Hub.
+    :param str event_hub_path: The path of the specific Event Hub to connect the client to.
+    :param credential: The credential object used for authentication which implements particular interface
+     of getting tokens. It accepts ~azure.eventhub.EventHubSharedKeyCredential,
+     ~azure.eventhub.EventHubSASTokenCredential, credential objects generated by the azure-identity library and
+     objects that implement get_token(self, *scopes) method.
+    :keyword bool network_tracing: Whether to output network trace logs to the logger. Default is `False`.
+    :keyword dict[str, Any] http_proxy: HTTP proxy settings. This must be a dictionary with the following
+     keys: 'proxy_hostname' (str value) and 'proxy_port' (int value).
+     Additionally the following keys may also be present: 'username', 'password'.
+    :keyword float auth_timeout: The time in seconds to wait for a token to be authorized by the service.
+     The default value is 60 seconds. If set to 0, no timeout will be enforced from the client.
+    :keyword str user_agent: The user agent that needs to be appended to the built in user agent string.
+    :keyword int retry_total: The total number of attempts to redo the failed operation when an error happened.
+     Default value is 3.
+    :keyword transport_type: The type of transport protocol that will be used for
+     communicating with the Event Hubs service. Default is ~azure.eventhub.TransportType.Amqp.
+        :class: ~azure.eventhub.TransportType
+
+    .. admonition:: Example:
+
+        .. literalinclude:: ../samples/sync_samples/sample_code_eventhub.py
+            :start-after: [START create_eventhub_producer_client_sync]
+            :end-before: [END create_eventhub_producer_client_sync]
+            :language: python
+            :dedent: 4
+            :caption: Create a new instance of the EventHubProducerClient.
+
+    """
+    def __init__(self, host, event_hub_path, credential, **kwargs):
+        # type:(str, str, Union[EventHubSharedKeyCredential, EventHubSASTokenCredential, TokenCredential], Any) -> None
+        super(EventHubProducerClient, self).__init__(
+            host=host, event_hub_path=event_hub_path, credential=credential, **kwargs)
+        self._producers = []  # type: List[EventHubProducer]
+        self._client_lock = threading.Lock()
+        self._producers_locks = []  # type: List[threading.Lock]
+        self._max_message_size_on_link = 0
+
+    def _init_locks_for_producers(self):
+        if not self._producers:
+            with self._client_lock:
+                if not self._producers:
+                    num_of_producers = len(self.get_partition_ids()) + 1
+                    self._producers = [None] * num_of_producers
+                    for _ in range(num_of_producers):
+                        self._producers_locks.append(threading.Lock())
```

Comment:
I'm wondering if it might be better to use a dictionary here rather than two lists with partition key indexes.....
Maybe something along the lines of:
```
self._producers = {pid: {'producer': None, 'lock': threading.Lock()} for pid in self.get_partition_ids}
```
```

Reply:
Non-blocking for now.